flight_sql/accessors/timestamp_array_accessor.cc

```diff
@@ -32,23 +34,36 @@ int64_t GetConversionToSecondsDivisor(TimeUnit::type unit) {
   return divisor;
 }
 
-uint32_t CalculateFraction(TimeUnit::type unit, uint64_t units_since_epoch) {
+uint32_t CalculateFraction(TimeUnit::type unit, int64_t units_since_epoch) {
   // Convert the given remainder and time unit to nanoseconds
   // since the fraction field on TIMESTAMP_STRUCT is in nanoseconds.
   switch (unit) {
   case TimeUnit::SECOND:
     return 0;
   case TimeUnit::MILLI:
     // 1000000 nanoseconds = 1 millisecond.
+    if (units_since_epoch < 0)
+      // See below regarding floor division; here we want ceiling division.
+      units_since_epoch += driver::odbcabstraction::MILLI_TO_SECONDS_DIVISOR *
+                           std::abs((units_since_epoch - (driver::odbcabstraction::MILLI_TO_SECONDS_DIVISOR - 1))
+                                    / driver::odbcabstraction::MILLI_TO_SECONDS_DIVISOR);
     return (units_since_epoch %
             driver::odbcabstraction::MILLI_TO_SECONDS_DIVISOR) *
            1000000;
   case TimeUnit::MICRO:
     // 1000 nanoseconds = 1 microsecond.
+    if (units_since_epoch < 0)
+      units_since_epoch += driver::odbcabstraction::MICRO_TO_SECONDS_DIVISOR *
+                           std::abs((units_since_epoch - (driver::odbcabstraction::MICRO_TO_SECONDS_DIVISOR - 1))
+                                    / driver::odbcabstraction::MICRO_TO_SECONDS_DIVISOR);
     return (units_since_epoch %
            driver::odbcabstraction::MICRO_TO_SECONDS_DIVISOR) * 1000;
   case TimeUnit::NANO:
     // 1000 nanoseconds = 1 microsecond.
```

Comment:
This comment is wrong (or at least irrelevant). Should say that fraction is already in nanoseconds.